xla/service/collective_quantizer.cc

```diff
@@ -76,49 +193,132 @@ bool IsSupportedQuantization(HloInstruction* instr, HloInstruction** convert,
-                                              ScalarBroadcast(scale_bcast)),
-                                          ScalarBroadcast()))) ||
-       Match(
-           ops.back(),
-           m::Convert(convert, m::Clamp(clamp, ScalarBroadcast(),
-                                        m::Divide(binary, m::Op(),
-                                                  ScalarBroadcast(scale_bcast)),
-                                        ScalarBroadcast()))))) {
-    unary_ops = {ops.begin(), ops.end() - 3};
-  } else if (ops.size() > 0 && Match(ops.back(), m::Convert(convert))) {
-    unary_ops = {ops.begin(), ops.end() - 1};
+             m::Convert(&subgraph.convert,
+                        m::Clamp(&subgraph.clamp, ScalarBroadcast(),
+                                 m::MultiplyAnyOrder(
+                                     &subgraph.binary, m::Op(),
+                                     ScalarBroadcast(&subgraph.scale_bcast)),
+                                 ScalarBroadcast()))) ||
+       Match(ops.back(),
+             m::Convert(
+                 &subgraph.convert,
+                 m::Clamp(&subgraph.clamp, ScalarBroadcast(),
+                          m::Divide(&subgraph.binary, m::Op(),
+                                    ScalarBroadcast(&subgraph.scale_bcast)),
+                          ScalarBroadcast()))))) {
+    subgraph.unaries = {ops.begin(), ops.end() - 3};
+  } else if (ops.size() > 0 &&
+             Match(ops.back(), m::Convert(&subgraph.convert))) {
+    subgraph.unaries = {ops.begin(), ops.end() - 1};
   } else {
     VLOG(5) << "Did not find type conversion or quantization pattern.";
-    return false;
+    return std::nullopt;
   }
 
   // The collected unary ops between collective and quantization/type conversion
   // may only include bitcast, copy, reshape and slice instructions.
-  for (HloInstruction* unary_op : unary_ops) {
-    if (!Match(unary_op, m::AnyOf<HloInstruction>(m::Bitcast(), m::Copy(),
-                                                  m::Reshape(), m::Slice()))) {
+  for (HloInstruction* unary : subgraph.unaries) {
+    if (!Match(unary, m::AnyOf<HloInstruction>(m::Bitcast(), m::Copy(),
+                                               m::Reshape(), m::Slice()))) {
       VLOG(5) << "Unexpected instruction in unary ops.";
-      return false;
+      return std::nullopt;
     }
   }
-  return true;
+  return std::make_optional<ConversionSubgraph>(subgraph);
 }
 
-bool IsSupportedCollective(HloInstruction* instr) {
-  return instr->opcode() == HloOpcode::kAllGather ||
-         instr->opcode() == HloOpcode::kAllToAll ||
-         instr->opcode() == HloOpcode::kCollectiveBroadcast ||
-         instr->opcode() == HloOpcode::kCollectivePermute;
+absl::Status MatchDequantization(HloInstruction* instr, bool* changed) {
+  std::optional<ConversionSubgraph> subgraph =
+      IsSupportedDequantization(instr->mutable_operand(0));
+
+  if (subgraph.has_value()) {
```

Comment:
Return early if the subgraph does not have a value. And similarly in MatchQuantization.